packages/yew/src/renderer.rs

```diff
@@ -0,0 +1,94 @@
+use std::cell::Cell;
+use std::panic::PanicInfo;
+use std::rc::Rc;
+
+use web_sys::Element;
+
+use crate::app_handle::AppHandle;
+use crate::html::IntoComponent;
+
+thread_local! {
+    static PANIC_HOOK_IS_SET: Cell<bool> = Cell::new(false);
+}
+
+/// Set a custom panic hook.
+/// Unless a panic hook is set through this function, Yew will
+/// overwrite any existing panic hook when one of the `start_app*` functions are called.
+#[cfg_attr(documenting, doc(cfg(feature = "render")))]
+pub fn set_custom_panic_hook(hook: Box<dyn Fn(&PanicInfo<'_>) + Sync + Send + 'static>) {
+    std::panic::set_hook(hook);
+    PANIC_HOOK_IS_SET.with(|hook_is_set| hook_is_set.set(true));
+}
+
+fn set_default_panic_hook() {
+    if !PANIC_HOOK_IS_SET.with(|hook_is_set| hook_is_set.replace(true)) {
+        std::panic::set_hook(Box::new(console_error_panic_hook::hook));
+    }
+}
+
+/// The Yew Renderer.
+///
+/// This is the main entry point of a Yew application.
+#[derive(Debug)]
+#[cfg_attr(documenting, doc(cfg(feature = "render")))]
+#[must_use = "Renderer does nothing unless render() is called."]
+pub struct Renderer<ICOMP>
+where
+    ICOMP: IntoComponent + 'static,
+{
+    root: Element,
+    props: ICOMP::Properties,
+}
+
+impl<ICOMP> Default for Renderer<ICOMP>
+where
+    ICOMP: IntoComponent + 'static,
+    ICOMP::Properties: Default,
+{
+    fn default() -> Self {
+        Self::with_props(Default::default())
+    }
+}
+
+impl<ICOMP> Renderer<ICOMP>
+where
+    ICOMP: IntoComponent + 'static,
+    ICOMP::Properties: Default,
+{
+    /// Creates a [Renderer] that renders into the document body with default properties.
+    pub fn new() -> Self {
```

Comment:
Considering there is a lot of differenT constructors below, maybe a builder pattern would be better here?

So that i could write like so:
```rust
yew::Renderer::<App>::new()
    .with_root(gloo::window())
    .with_props(some_props)
    .render();
```

Reply:
The current approach about `yew::Renderer::<App>::new().render()` implies that: `App::Properties` implements `Default` and for properties that does not implement default, it must use one of the `_with_props` constructor.

However, with a builder pattern, it means `render` can be called at anytime.

Consider the following code:

```rust

// App::Properties does not implement Default.
let renderer = yew::Renderer::<App>::new();

// we provide props here
let renderer = renderer.with_props(app_props);

// App is now rendered with app_props.
renderer.render();

```

```rust

// App::Properties does not implement Default.
let renderer = yew::Renderer::<App>::new();

// App does not have any available props to use, do we return `Result<AppHandle>` or panic?
renderer.render();

```

The following React-like approach is possible, but I think it's better to have a behaviour that is consistent with `ServerRenderer` in this pull request and maybe adjust the behaviour together in a future pull request?

```rust

let app = html!{
    <App />
};

let renderer = Renderer::new();
renderer.render(app);

```

Reply:
Im ok with either option.

The react like think you say we could add in future does look cool. Would be awesome if we did implement in the future.

Reply:
The idea of not "just" using `Html` to start the app is that `AppHandle` can provide `Scope` access to the top-level component so you can send messages to it from outside yew. I think a trait `IntoApplication` with `Html: IntoApplication` and `VChild<COMP>: IntoApplication` could unify the interface, where you'd use `html_nested!` to preserve the type of the top-level component as a `VChild` as usual.

Reply:
I guess this needs some design on how the React-like approach can work.

In general, for React components, most of time the higher component (or here, outside of the application) should impact the child component with props and avoid mutating child states directly. If you are having a pattern of parent attempts to mutates child components, you should consider moving the state to the parent. (i.e.: Instead of setting a new value on a text field (mutating state of child), the parent triggers a re-render by passing a new value as props to the text field and the value is stored in the parent. With a few notable exceptions, like `input.focus()`)

In this case, an updated layout can be re-rendered with a method on `AppHandle` with a different `html!` that provides a different props to the component instead of sending messages.

Since @voidpumpkin is fine with both options, let's keep the current behaviour here and improve it in a future pull request.

Reply:
Agreed, it could be a seperated follow-up PR.

Reply:
I also agree that the current behavior is fine. I'm not a fan of making behavior React-like (and introducing a macro call) when we have a way to express the call with the Rust type system. 

In this case, to provide a React-like API, we would be needlessly wrapping the type in VChild<_> and enforcing an opinionated pattern where the child *must* be updated through props (i.e. we would be taking away the ability to send messages to top-level component from outside Yew)

Reply:
Props-based approach also supports function components.

But let’s have this conversation another time.